dsc_lib/src/functions/substring.rs

```diff
@@ -0,0 +1,233 @@
+use rust_i18n::t;
+use serde_json::Value;
+use tracing::debug;
+
+#[derive(Debug, Default)]
+pub struct Substring {}
+
+impl Function for Substring {
+    fn get_metadata(&self) -> FunctionMetadata {
+        FunctionMetadata {
+            name: "substring".to_string(),
+            description: t!("functions.substring.description").to_string(),
+            category: vec![FunctionCategory::String],
+            min_args: 2,
+            max_args: 3,
+            accepted_arg_ordered_types: vec![
+                vec![FunctionArgKind::String],
+                vec![FunctionArgKind::Number],
+                vec![FunctionArgKind::Number],
+            ],
+            remaining_arg_accepted_types: None,
+            return_types: vec![FunctionArgKind::String],
+        }
+    }
+
+    fn invoke(&self, args: &[Value], _context: &Context) -> Result<Value, DscError> {
+        debug!("{}", t!("functions.substring.invoked"));
+
+        let string_to_parse = args[0].as_str().unwrap();
+        let start_index_value = args[1].as_i64().unwrap();
+
+        if start_index_value < 0 {
+            return Err(DscError::FunctionArg(
+                "substring".to_string(),
+                t!("functions.substring.startIndexNegative").to_string(),
+            ));
+        }
+
+        let start_index = usize::try_from(start_index_value).map_err(|_| {
+            DscError::FunctionArg(
+                "substring".to_string(),
+                t!("functions.substring.startIndexValueTooLarge").to_string(),
+            )
+        })?;
+        let string_length = string_to_parse.chars().count();
+
+        if start_index > string_length {
+            return Err(DscError::FunctionArg(
+                "substring".to_string(),
+                t!("functions.substring.startIndexTooLarge").to_string(),
+            ));
+        }
+
+        let length = if args.len() == 2 {
+            string_length - start_index
+        } else {
+            let length_value = args[2].as_i64().unwrap();
+
+            if length_value < 0 {
+                return Err(DscError::FunctionArg(
+                    "substring".to_string(),
+                    t!("functions.substring.lengthNegative").to_string(),
+                ));
+            }
+
+            let length = usize::try_from(length_value).map_err(|_| {
+                DscError::FunctionArg(
+                    "substring".to_string(),
+                    t!("functions.substring.lengthValueTooLarge").to_string(),
+                )
+            })?;
+
+            if start_index + length > string_length {
+                return Err(DscError::FunctionArg(
+                    "substring".to_string(),
+                    t!("functions.substring.lengthTooLarge").to_string(),
+                ));
+            }
+
+            length
+        };
+
+        let result: String = string_to_parse
+            .chars()
+            .skip(start_index)
+            .take(length)
+            .collect();
```

Comment:
I think it's simpler to do something like:

```suggestion
        let result: String = string_to_parse[start_index..length].to_string();
```

Reply:
I think that if I were to change to this, it would have the same implication as my earlier comment.